File: src/org/jgll/util/hashing/MurmurHash2.java
package org.jgll.util.hashing;

public class MurmurHash2 implements HashFunction{

	@Override
	public int hash(int a, int b, int c, int d) {
		final int m = 0x5bd1e995;
		final int r = 24;
		int h = a ^ 4;

		// b
		int k = b;
		k *= m;
		k ^= k >>> r;
		k *= m;

		h *= m;
		h ^= k;
		
		// c
		k = c;
		k *= m;
		k ^= k >>> r;
		k *= m;

		h *= m;
		h ^= k;
		
		// d
		k = d;
		k *= m;
		k ^= k >>> r;
		k *= m;

		h *= m;
		h ^= k;

		// last mix
		h *= m;
		h ^= h >>> 13;
		h *= m;
		h ^= h >>> 15;
		return h;
	}

	@Override
	public int hash(int a, int b, int c) {
		return hash(0, a, b, c);
	}

	@Override
	public int hash(int k) {
		return 0;
	}

	@Override
	public int hash(int k1, int k2) {
		return 0;
	}

}

File: src-scala/org/jgll_staged/util/hashing/hashfunction/MurmurHash2.scala
package org.jgll_staged.util.hashing.hashfunction

//remove if not needed
import scala.collection.JavaConversions._

@SerialVersionUID(1L)
class MurmurHash2(private var seed: Int) extends HashFunction {

  private val m = 0x5bd1e995

  private val r = 24

  def this() {
    this(0)
  }

  override def hash(a: Int, 
      b: Int, 
      c: Int, 
      d: Int): Int = {
    var h = seed ^ 4
    var k = a
    k = mixK(k)
    h = mixH(h, k)
    k = b
    k = mixK(k)
    h = mixH(h, k)
    k = c
    k = mixK(k)
    h = mixH(h, k)
    k = d
    k = mixK(k)
    h = mixH(h, k)
    h *= m
    h ^= h >>> 13
    h *= m
    h ^= h >>> 15
    h
  }

  private def mixK(_k: Int): Int = {
    var k = _k
    k *= m
    k ^= k >>> r
    k *= m
    k
  }

  private def mixH(_h: Int, k: Int): Int = {
    var h = _h
    h *= m
    h ^= k
    h
  }

  override def hash(a: Int, b: Int, c: Int): Int = hash(0, a, b, c)

  override def hash(k: Int): Int = 0

  override def hash(k1: Int, k2: Int): Int = 0

  override def hash(keys: Array[Int]): Int = 0

  override def hash(a: Int, 
      b: Int, 
      c: Int, 
      d: Int, 
      e: Int): Int = {
    var h = seed ^ 4
    var k = a
    k = mixK(k)
    h = mixH(h, k)
    k = b
    k = mixK(k)
    h = mixH(h, k)
    k = c
    k = mixK(k)
    h = mixH(h, k)
    k = d
    k = mixK(k)
    h = mixH(h, k)
    k = e
    k = mixK(k)
    h = mixH(h, k)
    h *= m
    h ^= h >>> 13
    h *= m
    h ^= h >>> 15
    h
  }
}
